From 0c7e7d5b92a141e997c989efd530259a3c586ccf Mon Sep 17 00:00:00 2001
From: dragonn <dragonn@op.pl>
Date: Wed, 2 Jul 2025 08:18:26 +0200
Subject: [PATCH 2/2] auto-brigthness

---
 drivers/platform/x86/asus-armoury.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/platform/x86/asus-armoury.c b/drivers/platform/x86/asus-armoury.c
index c5ce9d774c17..e7a3e1b5c39a 100644
--- a/drivers/platform/x86/asus-armoury.c
+++ b/drivers/platform/x86/asus-armoury.c
@@ -954,6 +954,7 @@ static const struct asus_attr_group armoury_attr_groups[] = {
 	{ &mcu_powersave_attr_group, ASUS_WMI_DEVID_MCU_POWERSAVE },
 	{ &panel_od_attr_group, ASUS_WMI_DEVID_PANEL_OD },
 	{ &panel_hd_mode_attr_group, ASUS_WMI_DEVID_PANEL_HD },
+	{ &screen_auto_brightness_attr_group, ASUS_WMI_DEVID_SCREEN_AUTO_BRIGHTNESS },
 };
 
 /**
-- 
2.50.0

